### PR TITLE
feat(motion): add createPresenceComponentVariant() factory

### PR DESCRIPTION
**Pull request by akatrukhin**
## New Behavior

PR adds `createPresenceComponentVariant()` that allows to customize `easing` & `duration` of a motion component created with `createPresenceComponent()`.

```ts
import { createPresenceComponentVariant } from '@fluentui/react-components';

const PresenceComponent = createPresenceComponent(/* --- */);
const PresenceVariant = createPresenceComponentVariant(PresenceComponent, {
  all: { duration: 500, easing: 'ease-in-out' },
});
```

### CODE REVIEW

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)

  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">39</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">33</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.regression.txt">Possible regression</a></td>
           </tr></table><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/Avatar-mount_base.html">622</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/Avatar-mount.html">650</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/Button-mount_base.html">312</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/Button-mount.html">292</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/Field-mount_base.html">1119</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/Field-mount.html">1077</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/FluentProvider-mount_base.html">695</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/FluentProvider-mount.html">722</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/FluentProviderWithTheme-mount_base.html">78</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/FluentProviderWithTheme-mount.html">81</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">39</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">33</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.regression.txt">Possible regression</a></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">76</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">78</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/MakeStyles-mount_base.html">874</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/MakeStyles-mount.html">845</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/Persona-mount_base.html">1763</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/Persona-mount.html">1716</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/SpinButton-mount_base.html">1334</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/SpinButton-mount.html">1397</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SwatchPicker</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/SwatchPicker-mount_base.html">1675</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/31943/perf-test-react-components/SwatchPicker-mount.html">1663</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Comment by fabricteam** [comment]
## 📊 Bundle size report

| Package & Exports | Baseline (minified/GZIP) | PR    | Change     |
| :---------------- | -----------------------: | ----: | ---------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ReactComponents.fixture.js'>react-components: entire library</abbr> | `1.095 MB`<br />`270.903 kB` | `1.095 MB`<br />`270.953 kB` | `81 B` <br />`50 B` |
| <samp>react-dialog</samp> <br /> <abbr title='bundle-size/Dialog.fixture.js'>Dialog (including children components)</abbr> | `99.002 kB`<br />`29.835 kB` | `99.089 kB`<br />`29.863 kB` | `87 B` <br />`28 B` |
| <samp>react-motion</samp> <br /> <abbr title='bundle-size/createPresenceComponent.fixture.js'>@fluentui/react-motion - createPresenceComponent()</abbr> | `4.846 kB`<br />`2.132 kB` | `4.925 kB`<br />`2.158 kB` | `79 B` <br />`26 B` |
| <samp>react-toast</samp> <br /> <abbr title='bundle-size/Toast.fixture.js'>Toast (including Toaster)</abbr> | `97.608 kB`<br />`29.358 kB` | `97.691 kB`<br />`29.389 kB` | `83 B` <br />`31 B` |

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports | Size (minified/GZIP) |
| ----------------- | -------------------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ButtonProviderAndTheme.fixture.js'>react-components: Button, FluentProvider & webLightTheme</abbr> | `69.141 kB`<br />`20.157 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `211.741 kB`<br />`60.957 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr> | `44.442 kB`<br />`14.607 kB` |
| <samp>react-motion</samp> <br /> <abbr title='bundle-size/createMotionComponent.fixture.js'>@fluentui/react-motion - createMotionComponent()</abbr> | `4.173 kB`<br />`1.847 kB` |
| <samp>react-motion</samp> <br /> <abbr title='bundle-size/PresenceGroup.fixture.js'>@fluentui/react-motion - PresenceGroup</abbr> | `1.714 kB`<br />`819 B` |
| <samp>react-portal-compat</samp> <br /> <abbr title='bundle-size/PortalCompat.fixture.js'>PortalCompatProvider</abbr> | `8.39 kB`<br />`2.64 kB` |
| <samp>react-timepicker-compat</samp> <br /> <abbr title='bundle-size/TimePicker.fixture.js'>TimePicker</abbr> | `106.611 kB`<br />`35.546 kB` |
</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/05ad86d34fef4aa2e2f2eb76182222014f07706f'>05ad86d34fef4aa2e2f2eb76182222014f07706f</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## 🕵 fluentuiv9 No visual regressions between this PR and main
<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

File: packages/react-components/react-motion/library/src/factories/createPresenceComponent.ts [hunk 1]
```diff
@@ -65,9 +65,23 @@ function shouldSkipAnimation(appear: boolean | undefined, isFirstMount: boolean,
   return !appear && isFirstMount && !!visible;
```
Comment (fabricteam) [inline packages/react-components/react-motion/library/src/factories/createPresenceComponent.ts:65]: ## [🕵 fluentuiv9 Open the Visual Regressions report to inspect the affected screenshots]( https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-approval/index.html?organization=uifabric&project=fabricpublic&prId=31943&commitId=7996e016b9c7516508ba6518d20f5a72972a49c1&env=prod&clientType=FLUENTUI)


  <details><summary><b>Avatar Converged</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Avatar Converged.badgeMask - RTL.normal.chromium.png|4| Changed|
|Avatar Converged.badgeMask.normal.chromium.png|4| Changed|

</details>


  <details><summary><b>Card Converged - Interactive</b> 3 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Card Converged - Interactive.appearance interactive - Filled - High Contrast.click.chromium.png|0| Removed|
|Card Converged - Interactive.appearance interactive - Filled - High Contrast.hover.chromium.png|0| Removed|
|Card Converged - Interactive.appearance interactive - Filled - High Contrast.normal.chromium.png|0| Removed|

</details>


  <details><summary><b>Image Converged</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Image Converged.Image Shape.normal.chromium.png|0| Removed|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

**Review by layershifter — APPROVED** [review]

**Pull request merged**